README.md
# loudgain

**loudgain** is a versatile ReplayGain 2.0 loudness normalizer, based on the EBU R128/ITU BS.1770 standard (-18 LUFS) and supports FLAC/Ogg/MP2/MP3/MP4/M4A/ALAC/Opus audio files. It uses the well-known `mp3gain` commandline syntax but will never modify the actual audio data.

Just what you ever wanted: The best of mp3gain, ReplayGain 2.0 and Linux combined. **Spread the word!**

_**Note:** There is a »runnable« version for 64-bit systems in the `bin` folder, called `loudgain.static`. Read [loudgain.static](#loudgainstatic).
You can use this if you only want to try out loudgain, cannot or don’t wish to compile it yourself. It starts much slower, but brings all its dependencies with it, and thus can even be used on older systems._

_**Mac users:** Read [Installation using Homebrew (Mac & Linux)](#installation-using-homebrew-mac--linux)._

---

## Table of Contents

<!-- MDTOC maxdepth:6 firsth1:2 numbering:0 flatten:0 bullets:1 updateOnSave:1 -->

- [Table of Contents](#table-of-contents)   
- [MAIN FEATURES](#main-features)   
- [NEWS, CHANGELOG](#news-changelog)   
- [GETTING STARTED](#getting-started)   
- [Installation using Homebrew (Mac & Linux)](#installation-using-homebrew-mac-linux)   
- [DEPENDENCIES](#dependencies)   
- [BUILDING](#building)   
- [TECHNICAL DETAILS (advanced users stuff)](#technical-details-advanced-users-stuff)   
   - [Things in the `bin`folder](#things-in-the-binfolder)   
      - [rgbpm – Folder-based loudness and BPM scanning](#rgbpmsh-–-folder-based-loudness-and-bpm-scanning)   
      - [loudgain.static](#loudgainstatic)   
   - [Uppercase or lowercase 'REPLAYGAIN_*' tags?](#uppercase-or-lowercase-replaygain_-tags)   
   - [Tags written (and/or deleted)](#tags-written-andor-deleted)   
      - [Default mode `-s i` (`--tag-mode i`)](#default-mode-s-i-tag-mode-i)   
      - [Enhanced mode `-s e` (`--tag-mode e`)](#enhanced-mode-s-e-tag-mode-e)   
      - ["LU" units mode `-s l` (`--tag-mode l`)](#lu-units-mode-s-l-tag-mode-l)   
   - [Track and Album Peak](#track-and-album-peak)   
   - [MP3 ID3v2.3, ID3v2.4, and APE tags](#mp3-id3v23-id3v24-and-ape-tags)   
      - [Force writing ID3v2.3 or ID3v2.4 tags](#force-writing-id3v23-or-id3v24-tags)   
      - [Strip unwanted ID3v1/APEv2 tags `-S` (`--striptags`)](#strip-unwanted-id3v1apev2-tags-s-striptags)   
   - [Analyze audio files and output to CSV](#analyze-audio-files-and-output-to-csv)   
      - [Example: Mary Black’s album »No Frontiers« (1989)](#example-mary-black’s-album-»no-frontiers«-1989)   
      - [Write a simple analysis to a CSV file](#write-a-simple-analysis-to-a-csv-file)   
      - [Analyze again, using clipping prevention](#analyze-again-using-clipping-prevention)   
      - [Just for fun: What if we _really_ followed the EBU R128 recommendation?](#just-for-fun-what-if-we-_really_-followed-the-ebu-r128-recommendation)   
      - [Time to save the tags and listen to the music!](#time-to-save-the-tags-and-listen-to-the-music)   
   - [Clipping prevention](#clipping-prevention)   
   - [Here be dragons: A word on MP3 files and ID3 tags](#here-be-dragons-a-word-on-mp3-files-and-id3-tags)   
      - [Some basic recommendations](#some-basic-recommendations)   
      - [A zillion ways to store loudness/gaining information](#a-zillion-ways-to-store-loudnessgaining-information)   
      - [Character encoding, or: Why does my smørrebrød look like "smÃ¸rrebrÃ¸d"?](#character-encoding-or-why-does-my-smørrebrød-look-like-smã¸rrebrã¸d)   
   - [How I handle Opus (.opus) audio files](#how-i-handle-opus-opus-audio-files)   
- [loudgain makes it easy following the »Gold Standard«](#loudgain-makes-it-easy-following-the-»gold-standard«)   
- [Quality over speed](#quality-over-speed)   
- [AUTHORS](#authors)   
- [COPYRIGHT](#copyright)   

<!-- /MDTOC -->

---

## MAIN FEATURES

**loudgain** is a loudness normalizer that scans music files and calculates
loudness-normalized gain and loudness peak values according to the EBU R128
standard, and can optionally write ReplayGain-compatible metadata.

[EBU R128](https://tech.ebu.ch/loudness) is a set of recommendations regarding
loudness normalisation based on the algorithms to measure audio loudness and
true-peak audio level defined in the
[ITU BS.1770](http://www.itu.int/rec/R-REC-BS.1770/en) standard, and is used in
the (currently under construction) [ReplayGain 2.0 specification](https://wiki.hydrogenaud.io/index.php?title=ReplayGain_2.0_specification).

loudgain implements a subset of mp3gain's command-line options, which means that
it can be used as a drop-in replacement in some situations.

**Note:** loudgain will _not_ modify the actual audio data, but instead just
write ReplayGain _tags_ if so requested. It is up to the player to interpret
these. (_Hint:_ In some players, you need to enable this feature.)

**Note:** loudgain can be used instead of `mp3gain`, `vorbisgain`, `metaflac`,
`aacgain` and others in order to write ReplayGain 2.0 compatible loudness tags
into MP2, MP3, Ogg Vorbis, FLAC, M4A/MP4 (AAC/ALAC audio) and Opus files, respectively.

**Note:** EBU R128 recommends a program (integrated) target loudness of -23 LUFS
and uses _LU_ and _LUFS_ units. The proposed ReplayGain 2.0 standard tries to
stay compatible with older software and thus uses EBU R128 loudness measuring
_but_ at a target of -18 LUFS (estimated to be equal to the old "89 dB" reference
loudness). The generated tags also still use the "dB" suffix (except for '-s l'
tag-mode which uses "LU"; 1 dB = 1 LU).

loudgain defaults to the ReplayGain 2.0 standard (-18 LUFS, "dB" units,
uppercase tags). Peak values are measured using the True Peak algorithm.
The only exception are Opus files which are 'nailed' to -23 LUFS by design.

**Standing on the shoulders of giants:** loudgain wouldn’t be possible without Linux and the fantastic [`FFmpeg`](https://ffmpeg.org/), [`taglib`](https://github.com/taglib/taglib) and [`libebur128`](https://github.com/jiixyj/libebur128) libraries. Thank you!

Also my heartfelt thanks to _Alessandro Ghedini_ who had the original idea back in 2014 and gave us his [v0.1 repo](https://github.com/ghedo/loudgain) from which I forked.

---

## NEWS, CHANGELOG

**2019-08-17** – **v0.6.1** released:
  * Can now show data for _.wav_ and other files (albeit not tag them). You could use that for a quick check or to write a (tab-separated) `.csv` file using `loudgain -O … > result.csv`. Or even with `awk` for further custom processing.
  * Fixed some channel layout and memory issues for files with more than 2 channels (i.e., 5.0, 5.1).

**2019-08-16** – **v0.6.0** released:
  * Rework the scanner module to work with the newer FFmpeg API (needed for FFmpeg 4.2+). Tested ok with FFmpeg 3.4.6, 4.1.4, 4.2.
  * Now officially provides the former `rgbpm.sh` as an updated and MacOS-compatible `rgbpm` command. After installing, type `rgbpm -h` for help. This is for _mass-tagging_ (1 album per folder).
  * Now provides a `[sudo] make uninstall` target, should you wish to uninstall it again.
  * Tested on Linux Mint 19.2, Ubuntu 18.04, Manjaro 18.0.4, MacOS 10.14.4 Mojave.

**2019-08-13** – **v0.5.4** released:
  * Make loudgain compile & run on MacOS X, with lots of help from @classicjazz.
  * Support for installation using [Homebrew](https://brew.sh/) (Apple) and [Linuxbrew](https://docs.brew.sh/Homebrew-on-Linux) (Linux).

**2019-08-06** – **v0.5.3** released:
  * Added _experimental (!)_ Opus (.opus) support. Please read [How I handle Opus (.opus) audio files](#how-i-handle-opus-opus-audio-files) and give feedback on [GitHub](https://github.com/Moonbase59/loudgain/issues)!

**2019-08-05** – **v0.5.2** released:
  * Add support for MPEG-1 Layer 2 (.mp2) audio files: Much like MP3, uses ID3v2 tags and the same options. These can be used with many taggers/players like _MusicBrainz Picard_, _IDJC_, _Audacious_ and many more.
  * `rgbpm.sh` also updated for `*.mp2` support.

**2019-08-02** – **v0.5.1** released:
  * Speedup: Eliminate unneccessary double file writes (deleting+rewriting RG tags). Should help with low-speed disks and network mounts.  
  * Error handling: loudgain now complains when unable to write to a file (instead of silently ignoring it).
  * Cosmetic changes: Update man page, help text and version info.

**2019-08-01** – **v0.5.0** released:
  * loudgain now uses FFmpeg’s `libswresample` instead of the deprecated `libavresample`. Fixes [#1](https://github.com/Moonbase59/loudgain/issues/1). Tested on Ubuntu 18.04, Linux Mint 19.1, Manjaro 18.0.4, Arch.

**2019-07-31** – **v0.4.1** released:
  * Added support for ALAC (Apple Lossless Audio Codec) audio in MPEG-4 containers
  (.m4a, .mp4, .alac). `-L` (`--lowercase`) option is supported.
  ReplayGain tags are written to `----:com.apple.iTunes:REPLAYGAIN_*`.
  * **Note:** I do not care about any »iTunes Soundcheck« (a.k.a »iTunNORM«) tags. If such should exist, they will stay untouched. Software outside Apple’s iTunes diaspora usually ignores these tags anyway.

**2019-07-31** – **v0.4.0** released:
  * Added support for AAC audio in MPEG-4 containers (.m4a, .mp4). `-L` (`--lowercase`) option is supported. ReplayGain tags are written to `----:com.apple.iTunes:REPLAYGAIN_*`.
  * loudgain was never intended for video files, but it actually _does_ work on .mp4 video files, although only one audio track is respected. Go figure!

**2019-07-17** — **v0.3.2** released:
  * Work around [bug #913 in taglib](https://github.com/taglib/taglib/issues/913): If both APEv2 and ID3v1 tags present and due to be stripped, sometimes ID3v1 tags would be left over.

**2019-07-17** — **v0.3.1** released:
  * "Delete tags" `-s d` (`--tagmode=d`) now respects "strip" `-S` (`--striptags`) and "ID3v2 version" `-I` (`--id3v2version`) settings for MP3 files. Thus, when removing ReplayGain tags, unwanted tag types like ID3v1 or APE can also be removed, and the the resultant ID3v2 tag version be specified.

**2019-07-13** — **v0.3.0** released:
  * Much better clipping prevention logic.
  * Breaking change: Some long options slightly renamed.
  * Breaking change: `-k` now obeys EBU -1 dBTP maximum true peak level (as opposed to 0 dBTP before). Read [Clipping prevention](#clipping-prevention).
  * New option `-K n` (`--maxtpl=n`) switches on clipping prevention like `-k` and allows manually setting the max. true peak level (i.e., `-K 0` to limit at digital full scale, or `-K -2` if later processing includes lossy encoding).
  * Breaking change: Reordered and extended `-O` (`--output-new`) output. It now adds the fields `New_Peak` and `New_Peak_dBTP` so one can see what will happen in the playout/post-processing chain after the replay gain has been applied. Read [Analyze audio files and output to CSV](#analyze-audio-files-and-output-to-csv).

**2019-07-12** — Temporary solution to the taglib problem:

debian/changelog
loudgain (0.5.4-1) unstable; urgency=medium

  * Version 0.5.4

 -- Matthias C. Hormann <mhormann@gmx.de>  Tue, 13 Aug 2019 06:40:09 +0200

loudgain (0.5.3-1ubuntu1) unstable; urgency=medium

  * Version 0.5.3

 -- Matthias C. Hormann <mhormann@gmx.de>  Fri, 09 Aug 2019 18:22:23 +0200

loudgain (0.1-1) unstable; urgency=medium

  * Initial release (Closes: #nnnn)

 -- Alessandro Ghedini <ghedo@debian.org>  Mon, 26 Jun 2017 22:10:59 +0100


src/loudgain.c
/*
 * Loudness normalizer based on the EBU R128 standard
 *
 * Copyright (c) 2014, Alessandro Ghedini
 * All rights reserved.
 *
 * 2019-06-30 - v0.2.1 - Matthias C. Hormann
 *  - Added version
 *  - Added writing tags to Ogg Vorbis files (now supports MP3, FLAC, Ogg Vorbis)
 *  - Always remove REPLAYGAIN_REFERENCE_LOUDNESS, wrong value might confuse players
 *  - Added notice in help on which file types can be written
 *  - Added album summary
 * 2019-07-07 - v0.2.2 - Matthias C. Hormann
 *  - Fixed album peak calculation.
 *  - Write REPLAYGAIN_ALBUM_* tags only if in album mode
 *  - Better versioning (CMakeLists.txt → config.h)
 *  - TODO: clipping calculation still wrong
 * 2019-07-08 - v0.2.4 - Matthias C. Hormann
 *  - add "-s e" mode, writes extra tags (REPLAYGAIN_REFERENCE_LOUDNESS,
 *    REPLAYGAIN_TRACK_RANGE and REPLAYGAIN_ALBUM_RANGE)
 *  - add "-s l" mode (like "-s e" but uses LU/LUFS instead of dB)
 * 2019-07-08 - v0.2.5 - Matthias C. Hormann
 *  - Clipping warning & prevention (-k) now works correctly, both track & album
 * 2019-07-09 - v0.2.6 - Matthias C. Hormann
 *  - Add "-L" mode to force lowercase tags in MP3/ID3v2.
 * 2019-07-10 - v0.2.7 - Matthias C. Hormann
 *  - Add "-S" mode to strip ID3v1/APEv2 tags from MP3 files.
 *  - Add "-I 3"/"-I 4" modes to select ID3v2 version to write.
 *  - First step to implement a new tab-delimited list format: "-O" mode.
 * 2019-07-13 - v0.2.8 - Matthias C. Hormann
 *  - new -O output format: re-ordered, now shows peak before/after gain applied
 *  - -k now defaults to clipping prevention at -1 dBTP (as EBU recommends)
 *  - New -K: Allows clippping prevention with settable dBTP level,
 *     i.e. -K 0 (old-style) or -K -2 (to compensate for post-processing losses)
 * 2019-08-06 - v0.5.3 - Matthias C. Hormann
 *  - Add support for Opus (.opus) files.
 * 2019-08-16 - v0.6.0 - Matthias C. Hormann
 *  - Rework for new FFmpeg API (get rid of deprecated calls)
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#include <math.h>
#include <getopt.h>

#include <ebur128.h>
#include <libavcodec/avcodec.h>
#include <libavutil/common.h>
#include <libswresample/swresample.h>
#include <libavformat/avformat.h>


#include "config.h"
#include "scan.h"
#include "tag.h"
#include "printf.h"

const char *short_opts = "rackK:d:oOqs:LSI:h?v";

static struct option long_opts[] = {
	{ "track",        no_argument,       NULL, 'r' },
	{ "album",        no_argument,       NULL, 'a' },

	{ "clip",         no_argument,       NULL, 'c' },
	{ "noclip",       no_argument,       NULL, 'k' },
	{ "maxtpl",       required_argument, NULL, 'K' },

	{ "pregain",      required_argument, NULL, 'd' },

	{ "output",       no_argument,       NULL, 'o' },
	{ "output-new",   no_argument,       NULL, 'O' },
	{ "quiet",        no_argument,       NULL, 'q' },

	{ "tagmode",      required_argument, NULL, 's' },
	{ "lowercase",    no_argument,       NULL, 'L' },
	{ "striptags",    no_argument,       NULL, 'S' },
	{ "id3v2version", required_argument, NULL, 'I' },

	{ "help",         no_argument,       NULL, 'h' },
	{ "version",      no_argument,       NULL, 'v' },
	{ 0, 0, 0, 0 }
};

bool warn_ebu            = false;
int  ebur128_v_major     = 0;
int  ebur128_v_minor     = 0;
int  ebur128_v_patch     = 0;
char ebur128_version[15] = "";
unsigned swr_ver         = 0;
char     swr_version[15] = "";
unsigned lavf_ver         = 0;
char     lavf_version[15] = "";

static inline void help(void);
static inline void version(void);

int main(int argc, char *argv[]) {
	int rc, i;

	char mode           = 's';
	char unit[3]        = "dB";

	unsigned nb_files   = 0;

	double pre_gain     = 0.f;
	double max_true_peak_level = -1.0; // dBTP; default for -k, as per EBU Tech 3343

	bool no_clip        = false;
	bool warn_clip      = true;
	bool do_album       = false;
	bool tab_output     = false;
	bool tab_output_new = false;
	bool lowercase      = false; // force MP3 ID3v2 tags to lowercase?
	bool strip          = false; // MP3 ID3v2: strip other tag types?
	int  id3v2version   = 4;     // MP3 ID3v2 version to write; can be 3 or 4

	// libebur128 version check -- versions before 1.2.4 aren’t recommended
	ebur128_get_version(&ebur128_v_major, &ebur128_v_minor, &ebur128_v_patch);
	snprintf(ebur128_version, sizeof(ebur128_version), "%d.%d.%d",
		ebur128_v_major, ebur128_v_minor, ebur128_v_patch);
	if (ebur128_v_major <= 1 && ebur128_v_minor <= 2 && ebur128_v_patch < 4)
		warn_ebu = true;

	// libavformat version
	lavf_ver = avformat_version();
	snprintf(lavf_version, sizeof(lavf_version), "%u.%u.%u",
		lavf_ver>>16, lavf_ver>>8&0xff, lavf_ver&0xff);

	// libswresample version
	swr_ver = swresample_version();
	snprintf(swr_version, sizeof(swr_version), "%u.%u.%u",
		swr_ver>>16, swr_ver>>8&0xff, swr_ver&0xff);

	while ((rc = getopt_long(argc, argv, short_opts, long_opts, &i)) !=-1) {
		switch (rc) {
			case 'r':
				/* noop */
				break;

			case 'a':
				do_album = true;
				break;

			case 'c':
				warn_clip = false;
				break;

			case 'k':
				// old-style, no argument, now defaults to -1 dBTP max. true peak level
				no_clip = true;
				break;

			case 'K': {
				// new style, argument in dBTP, sets max. true peak level
				no_clip = true;

				char *rest = NULL;
				max_true_peak_level = strtod(optarg, &rest);

				if (!rest ||
				    (rest == optarg) ||
				    !isfinite(pre_gain))
					fail_printf("Invalid max. true peak level (dBTP)");
				break;
			}

			case 'd': {
				char *rest = NULL;
				pre_gain = strtod(optarg, &rest);

				if (!rest ||
				    (rest == optarg) ||
				    !isfinite(pre_gain))
					fail_printf("Invalid pregain value (dB/LU)");
				break;
			}

			case 'o':
				tab_output = true;
				break;

			case 'O':
				tab_output_new = true;
				break;

			case 'q':
				quiet = 1;
				break;

			case 's': {
				// for mp3gain compatibilty, include modes that do nothing
				char *valid_modes = "cdielavsr";
				mode = optarg[0];
				if (strchr(valid_modes, mode) == NULL)
					fail_printf("Invalid tag mode: '%c'", mode);
				if (mode == 'l') {
					strcpy(unit, "LU");
				}
				break;
			}

			case 'L':
				lowercase = true;
				break;

			case 'S':
				strip = true;
				break;

			case 'I':
				id3v2version = atoi(optarg);
				if (!(id3v2version == 3) && !(id3v2version == 4))
					fail_printf("Invalid ID3v2 version; only 3 and 4 are supported.");
				break;

			case '?':
				if (optopt == 0) {
					// actual option '-?'
					help();
					return 0;
				} else {
					// getopt error, message already printed
					return 1;	// error
				}
			case 'h':
				help();
				return 0;

			case 'v':
				version();
				return 0;
		}
	}

	nb_files = argc - optind;

	scan_init(nb_files);

	for (i = optind; i < argc; i++) {
		ok_printf("Scanning '%s' ...", argv[i]);

		scan_file(argv[i], i - optind);
	}

	if (tab_output)
		printf("File\tMP3 gain\tdB gain\tMax Amplitude\tMax global_gain\tMin global_gain\n");

	if (tab_output_new)
		printf("File\tLoudness\tRange\tTrue_Peak\tTrue_Peak_dBTP\tReference\tWill_clip\tClip_prevent\tGain\tNew_Peak\tNew_Peak_dBTP\n");

	for (i = 0; i < nb_files; i++) {
		bool will_clip = false;
		double tgain = 1.0; // "gained" track peak
		double tnew;
		double tpeak = pow(10.0, max_true_peak_level / 20.0); // track peak limit
		double again = 1.0; // "gained" album peak
		double anew;
		double apeak = pow(10.0, max_true_peak_level / 20.0); // album peak limit
		bool tclip = false;
		bool aclip = false;

		scan_result *scan = scan_get_track_result(i, pre_gain);

		if (scan == NULL)
			continue;

		if (do_album)
			scan_set_album_result(scan, pre_gain);

		// Check if track or album will clip, and correct if so requested (-k/-K)

		// track peak after gain
		tgain = pow(10.0, scan -> track_gain / 20.0) * scan -> track_peak;
		tnew = tgain;
		if (do_album) {
			// album peak after gain
			again = pow(10.0, scan -> album_gain / 20.0) * scan -> album_peak;
			anew = again;
		}

		if ((tgain > tpeak) || (do_album && (again > apeak)))
			will_clip = true;

		// printf("\ntrack: %.2f LU, peak %.6f; album: %.2f LU, peak %.6f\ntrack: %.6f, %.6f; album: %.6f, %.6f; Clip: %s\n",
		// 	scan -> track_gain, scan -> track_peak, scan -> album_gain, scan -> album_peak,
		// 	tgain, tpeak, again, apeak, will_clip ? "Yes" : "No");

		if (will_clip && no_clip) {
			if (tgain > tpeak) {
				// set new track peak = minimum of peak after gain and peak limit
				tnew = FFMIN(tgain, tpeak);
				scan -> track_gain = scan -> track_gain - (log10(tgain/tnew) * 20.0);
				tclip = true;
			}

			if (do_album && (again > apeak)) {
				anew = FFMIN(again, apeak);
				scan -> album_gain = scan -> album_gain - (log10(again/anew) * 20.0);
				aclip = true;
			}

			will_clip = false;

			// printf("\nAfter clipping prevention:\ntrack: %.2f LU, peak %.6f; album: %.2f LU, peak %.6f\ntrack: %.6f, %.6f; album: %.6f, %.6f; Clip: %s\n",
			// 	scan -> track_gain, scan -> track_peak, scan -> album_gain, scan -> album_peak,
			// 	tgain, tpeak, again, apeak, will_clip ? "Yes" : "No");
		}

		switch (mode) {
			case 'c': /* check tags */
				break;

			case 'd': /* delete tags */
				switch (scan -> codec_id) {
					case AV_CODEC_ID_MP2:
					case AV_CODEC_ID_MP3:
						if (!tag_clear_mp3(scan, strip, id3v2version))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					case AV_CODEC_ID_FLAC:
						if (!tag_clear_flac(scan))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					case AV_CODEC_ID_VORBIS:
						if (!tag_clear_vorbis(scan))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					case AV_CODEC_ID_AAC:
					case AV_CODEC_ID_ALAC:
						if (!tag_clear_mp4(scan))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					case AV_CODEC_ID_OPUS:
						if (!tag_clear_opus(scan))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					default:
						err_printf("File type not supported");
						break;
				}
				break;

			case 'i': /* ID3v2 tags */
			case 'e': /* same as 'i' plus extra tags */
			case 'l': /* same as 'e' but in LU/LUFS units (instead of 'dB')*/
				switch (scan -> codec_id) {
					case AV_CODEC_ID_MP2:
					case AV_CODEC_ID_MP3:
						// tag_clear_mp3(scan, strip, id3v2version);
						if (!tag_write_mp3(scan, do_album, mode, unit, lowercase, strip, id3v2version))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					case AV_CODEC_ID_FLAC:
						// tag_clear_flac(scan);
						if (!tag_write_flac(scan, do_album, mode, unit))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					case AV_CODEC_ID_VORBIS:
						// tag_clear_vorbis(scan);
						if (!tag_write_vorbis(scan, do_album, mode, unit))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					case AV_CODEC_ID_AAC:
					case AV_CODEC_ID_ALAC:
						// tag_clear_mp4(scan);
						if (!tag_write_mp4(scan, do_album, mode, unit, lowercase))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					case AV_CODEC_ID_OPUS:
						// tag_clear_opus(scan);
						if (!tag_write_opus(scan, do_album, mode, unit))
							err_printf("Couldn't write to: %s", scan -> file);
						break;

					default:
						err_printf("File type not supported");
						break;
				}
				break;

			case 'a': /* APEv2 tags */
				err_printf("APEv2 tags are not supported");
				break;

			case 'v': /* Vorbis Comments tags */
				err_printf("Vorbis Comment tags are not supported");
				break;

			case 's': /* skip tags */
				break;

			case 'r': /* force re-calculation */
				break;

			default:
				err_printf("Invalid tag mode");
				break;
		}

		if (tab_output) {
			// output old-style mp3gain-compatible list
			printf("%s\t", scan -> file);
			printf("%d\t", 0);
			printf("%.2f\t", scan -> track_gain);
			printf("%.6f\t", scan -> track_peak * 32768.0);
			printf("%d\t", 0);
			printf("%d\n", 0);

			if (warn_clip && will_clip)
				err_printf("The track will clip");

			if ((i == (nb_files - 1)) && do_album) {
				printf("%s\t", "Album");
				printf("%d\t", 0);
				printf("%.2f\t", scan -> album_gain);
				printf("%.6f\t", scan -> album_peak * 32768.0);
				printf("%d\t", 0);
				printf("%d\n", 0);
			}
		} else if (tab_output_new) {
			// output new style list: File;Loudness;Range;Gain;Reference;Peak;Peak dBTP;Clipping;Clip-prevent
			printf("%s\t", scan -> file);
			printf("%.2f LUFS\t", scan -> track_loudness);
			printf("%.2f %s\t", scan -> track_loudness_range, unit);
			printf("%.6f\t", scan -> track_peak);
			printf("%.2f dBTP\t", 20.0 * log10(scan -> track_peak));
			printf("%.2f LUFS\t", scan -> loudness_reference);
			printf("%s\t", will_clip ? "Y" : "N");
			printf("%s\t", tclip ? "Y" : "N");
			printf("%.2f %s\t", scan -> track_gain, unit);
			printf("%.6f\t", tnew);
			printf("%.2f dBTP\n", 20.0 * log10(tnew));

			if ((i == (nb_files - 1)) && do_album) {
				printf("%s\t", "Album");
				printf("%.2f LUFS\t", scan -> album_loudness);
				printf("%.2f %s\t", scan -> album_loudness_range, unit);
				printf("%.6f\t", scan -> album_peak);
				printf("%.2f dBTP\t", 20.0 * log10(scan -> album_peak));
				printf("%.2f LUFS\t", scan -> loudness_reference);
				printf("%s\t", (!aclip && (again > apeak)) ? "Y" : "N");
				printf("%s\t", aclip ? "Y" : "N");
				printf("%.2f %s\t", scan -> album_gain, unit);
				printf("%.6f\t", anew);
				printf("%.2f dBTP\n", 20.0 * log10(anew));
			}
		} else {
			// output something human-readable
			printf("\nTrack: %s\n", scan -> file);

			printf(" Loudness: %8.2f LUFS\n", scan -> track_loudness);
			printf(" Range:    %8.2f %s\n", scan -> track_loudness_range, unit);
			printf(" Peak:     %8.6f (%.2f dBTP)\n", scan -> track_peak, 20.0 * log10(scan -> track_peak));
			if (scan -> codec_id == AV_CODEC_ID_OPUS) {
				// also show the Q7.8 number that goes into R128_TRACK_GAIN
				printf(" Gain:     %8.2f %s (%d)%s\n", scan -> track_gain, unit,
				 gain_to_q78num(scan -> track_gain),
				 tclip ? " (corrected to prevent clipping)" : "");
			} else {
				printf(" Gain:     %8.2f %s%s\n", scan -> track_gain, unit,
				 tclip ? " (corrected to prevent clipping)" : "");
			}

			if (warn_clip && will_clip)
				err_printf("The track will clip");

			if ((i == (nb_files - 1)) && do_album) {
				printf("\nAlbum:\n");

				printf(" Loudness: %8.2f LUFS\n", scan -> album_loudness);
				printf(" Range:    %8.2f %s\n", scan -> album_loudness_range, unit);
				printf(" Peak:     %8.6f (%.2f dBTP)\n", scan -> album_peak, 20.0 * log10(scan -> album_peak));
				if (scan -> codec_id == AV_CODEC_ID_OPUS) {
					// also show the Q7.8 number that goes into R128_ALBUM_GAIN
					printf(" Gain:     %8.2f %s (%d)%s\n", scan -> album_gain, unit,
					gain_to_q78num(scan -> album_gain),
						aclip ? " (corrected to prevent clipping)" : "");
				} else {
					printf(" Gain:     %8.2f %s%s\n", scan -> album_gain, unit,
						aclip ? " (corrected to prevent clipping)" : "");
				}
			}
		}

		free(scan);
	}

	scan_deinit();

	return 0;
}

static inline void help(void) {
	#define CMD_HELP(CMDL, CMDS, MSG) printf("  %s%-5s %-16s%s  %s.\n", COLOR_YELLOW, CMDS ",", CMDL, COLOR_OFF, MSG);

	printf(COLOR_RED "Usage: " COLOR_OFF);
	printf("%s%s%s ", COLOR_GREEN, PROJECT_NAME, COLOR_OFF);
	puts("[OPTIONS] FILES...\n");

	printf("%s %s supports writing tags to the following file types:\n", PROJECT_NAME, PROJECT_VER);
	puts("  FLAC (.flac), Ogg Vorbis (.ogg), MP2 (.mp2), MP3 (.mp3), MP4 (.mp4, .m4a).");
	puts("  Opus (.opus) -- experimental support, use with care!\n");

	if (warn_ebu) {
		printf("%sWarning:%s Your EBU R128 library (libebur128) is version %s.\n", COLOR_RED, COLOR_OFF, ebur128_version);
		puts("This is an old version and might cause problems.\n");
	}

	puts(COLOR_RED "Options:" COLOR_OFF);

	CMD_HELP("--help",     "-h", "Show this help");
	CMD_HELP("--version",  "-v", "Show version number");

	puts("");

	CMD_HELP("--track",  "-r", "Calculate track gain only (default)");
	CMD_HELP("--album",  "-a", "Calculate album gain (and track gain)");

	puts("");

	CMD_HELP("--clip",   "-c", "Ignore clipping warning");
	CMD_HELP("--noclip", "-k", "Lower track/album gain to avoid clipping (<= -1 dBTP)");
	CMD_HELP("--maxtpl=n", "-K n", "Avoid clipping; max. true peak level = n dBTP");

	CMD_HELP("--pregain=n",  "-d n",  "Apply n dB/LU pre-gain value (-5 for -23 LUFS target)");

	puts("");

	CMD_HELP("--tagmode=d", "-s d",  "Delete ReplayGain tags from files");
	CMD_HELP("--tagmode=i", "-s i",  "Write ReplayGain 2.0 tags to files");
	CMD_HELP("--tagmode=e", "-s e",  "like '-s i', plus extra tags (reference, ranges)");
	CMD_HELP("--tagmode=l", "-s l",  "like '-s e', but LU units instead of dB");

	CMD_HELP("--tagmode=s", "-s s",  "Don't write ReplayGain tags (default)");

	puts("");

	CMD_HELP("--lowercase", "-L", "Force lowercase tags (MP2/MP3/MP4 only; non-standard)");
	CMD_HELP("--striptags", "-S", "Strip tag types other than ID3v2 from MP2/MP3");
	CMD_HELP("--id3v2version=3", "-I 3", "Write ID3v2.3 tags to MP2/MP3 files");
	CMD_HELP("--id3v2version=4", "-I 4", "Write ID3v2.4 tags to MP2/MP3 files (default)");

	puts("");

	CMD_HELP("--output",     "-o",  "Database-friendly tab-delimited list output");
	CMD_HELP("--output-new", "-O",  "New format tab-delimited list output");
	CMD_HELP("--quiet",      "-q",  "Don't print scanning status messages");

	puts("");
	// puts("Mandatory arguments to long options are also mandatory for any corresponding short options.");
	// puts("");
}

static inline void version(void) {
	printf("%s %s - using:\n", PROJECT_NAME, PROJECT_VER);
	printf("  %s %s\n", "libebur128", ebur128_version);
	printf("  %s %s\n", "libavformat", lavf_version);
	printf("  %s %s\n", "libswresample", swr_version);
}
